### DISCUSSION TITLE
Binding to ISourceList

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#8331] [Q&A]
I have Sources that contains FileItem and also FolderItem that contain various FileItem.

I want to display "x files left" the Reactive way (I could do it another way, but want to practice with this simple scenario).

I changed Sources from ObservableCollection to SourceList (plus I read that ObservableCollection is not thread-safe and I'm doing multi-thread processing).

First problem: How can I find the list to the ListBox? This doesn't work `Items="{Binding Encoder.Sources.Items}"`

Then... how can I obtain how many files remain (direct files and files within folders)? I want to bind it to the UI as "x files left". I got this so far.

```
var combined = new SourceList<IObservable<int>>();
var sourceObserve = Encoder.Sources.Connect();
combined.Add(sourceObserve.Filter(x => x is not FolderItem).Count());
combined.Add(SumEx.Sum(sourceObserve.Filter(x => x is FolderItem), x =>((FolderItem)x).Files.Count));
```

**Comment by maxkatz6** [comment]
It seems like ReactiveUI recommends to use their special implementation of obs.collection - ReadOnlyObservableCollection. And that's what I was doing before when working with DynamicData and SourceLists. See their "how to use" docs:
https://www.reactiveui.net/docs/handbook/collections/

You basically need Connect().Bind() methods combined. 

Not sure if I understand second question.

**Comment by mysteryx93** [comment]
2nd question. I have Encoder.Sources that contains File1, File2, File3, Folder1 containing File4, File5 and File6. That means Encoder.Sources.Count = 4, and Encoder.Sources[3].Files.Count = 3.

I want to create a Reactive bindable property that says "6 files left". How?

**Reply by maxkatz6** [reply]
Easiest way would be something like this (untested):
```csharp
Encoder.Sources.Connect().ObserveOn().Bind(out  var materializedObservableCollection);  // show this collection on the UI

IObservable<int> countObservable = Encoder.Sources.Connect()
    .Select(_ => materializedObservableCollection.Select(item => item switch
    {
        File => 1,
        Folder folder => folder.Files.Count
    }).Sum());
```
Where we have already materialized collection to show on UI, so we can reuse it to calculate "files left".
`IObservable<int>` then can be bind to the UI directly or with some property in between (assign to a reactive prop inside of Subscibe method or use ObservableAsPropertyHelper). Don't forget about ObserveOn to avoid changes of UI bounded properties from the invalid thread.

`.Connect()` method returns observable change sets which can be used to calculate total left based on difference. I.e. `initial +  getLeft(changeSet1) + getLeft(changeSet2) + getLeft(changeSet3)` where `getLeft` would work only on IChangeSet. In this way you don't need to have any materialized collections.

**Reply by maxkatz6** [reply]
But I am not an ReactiveUI expert, and this isn't a ReactiveUI repository. So, it probably not the best possible answer.

**Reply by mysteryx93** [reply]
wow I would never have thought of doing it this way. For some reason I'm having a very hard time starting to think the "Reactive" way. I'll probably just stick to ObservableCollection for Encoder.Sources (can use ToObservableChangeSet instead); then how does that impact `materializedObservableCollection.Select`?

Will your code respond to changes to the list of files within a folder? Because it's mostly the parent list we're tracking.
